### ISSUE TITLE
Auto detect resource info

### ISSUE THREAD

**Issue by james-bebbington** [open-telemetry/opentelemetry-go#785]
**Background:**
Resource information can currently be added to tracer / meter providers or added in an exporter. It would be useful to have the option to automatically detect resource information from the host (e.g. from an environment variable or from aws, gcp, etc metadata) and use this rather than manually configuring resource information.

**Proposal:**
Implement a simple ResourceDetector interface (including a default FromEnv implementation), with a function to allow adding custom detectors - custom detectors (which would include vendor specific code) can then live as separate modules in Contrib repos (and could be re-used by the Collector), e.g.

```go
type ResourceDetector interface {
    Detect(ctx context.Context) (*resource.Resource, error)
}

func AddDetectors(detector ...ResourceDetector) { ... }

// for each added ResourceDetector, call Detect() & return resolved resource
func AutoDetect(ctx context.Context) (*resource.Resource, error) { ... }
```

Usage:
```go
resourceprovider.AddDetectors(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)
resource := resourceprovider.AutoDetect(ctx)
push.New(selector, exporter, push.WithResource(resource))
```

**Related Work:**

- This is somewhat similar to what was done in the OpenCensus Go SDK [here](https://github.com/census-instrumentation/opencensus-go/blob/master/resource/resource.go) & [here](https://github.com/census-ecosystem/opencensus-go-resource)
- See similar implementation in OpenTelemetry JS SDK [here](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-resources)

**Comment by james-bebbington** [comment]
Note I am actually currently more interested in implementing this for the Collector, but expect that it would make more sense for this to live in the SDK so that custom detector code can be shared.

See similar issue in Collector repo: https://github.com/open-telemetry/opentelemetry-collector/issues/871
And related draft PR for discussion: https://github.com/open-telemetry/opentelemetry-collector/pull/1059

**Comment by nilebox** [comment]
Note that If we decide to implement this feature in the Go SDK, it should be consistently ported to other languages as well.

It would probably also make sense to document this as a language agnostic requirement in the spec then.

**Comment by james-bebbington** [comment]
> Note that If we decide to implement this feature in the Go SDK, it should be consistently ported to other languages as well.
> 
> It would probably also make sense to document this as a language agnostic requirement in the spec then.

Yes good point. If we go with this proposal, then [this doc](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/sdk.md) should presumably be updated. In hindsight, maybe I should have created this issue in the specs repo first?

**Comment by jrcamp** [comment]
@james-bebbington I've commented on https://github.com/open-telemetry/oteps/pull/111, I assume we should focus our discussions there since there's multiple PR's and issues involved?

**Comment by james-bebbington** [comment]
Yes sorry I meant to comment here earlier. I moved this to an OTEP so feel free to close this issue in the meantime.

**Comment by MrAlias** [comment]
Closed by #939
